python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -1723,6 +1748,130 @@ def emptyTrash(self, ignore_errors: bool = True) -> None:
+        # associated with these refs. Or derived ones if they don't exist
+        # in the source.
+        source_records = source_datastore._get_stored_records_associated_with_refs(refs)
+
+        source_ids = set(source_records)
+        log.debug("Number of datastore records find in source: %d", len(source_ids))
+
+        # The not None check is to appease mypy
+        requested_ids = set(ref.id for ref in refs if ref.id is not None)
+        missing_ids = requested_ids - source_ids
+
+        # Missing IDs can be okay if that datastore has allowed
+        # gets based on file existence. Should we transfer what we can
+        # or complain about it and warn?
+        if missing_ids and not source_datastore.trustGetRequest:
+            raise ValueError(f"Some datasets are missing from source datastore {source_datastore}:"
+                             f" {missing_ids}")
+
+        # Need to map these missing IDs to a DatasetRef so we can guess
+        # the details.
+        if missing_ids:
+            log.info("Number of expected datasets missing from source datastore records: %d",
+                     len(missing_ids))
+            id_to_ref = {ref.id: ref for ref in refs if ref.id in missing_ids}
+
+            for missing in missing_ids:
+                expected = self._get_expected_dataset_locations_info(id_to_ref[missing])
+                source_records[missing].extend(info for _, info in expected)
+
+        # See if we already have these records
+        target_records = self._get_stored_records_associated_with_refs(target_refs)
+
+        # The artifacts to register
+        artifacts = []
+
+        # Refs that already exist
+        already_present = []
+
+        # Now can transfer the artifacts
+        for source_ref, target_ref in zip(refs, target_refs):
+            if target_ref.id in target_records:
+                # Already have an artifact for this.
+                already_present.append(target_ref)
+                continue
+
+            # mypy needs to know these are always resolved refs
+            assert source_ref.id is not None, "Unexpectedly got unresolved ref inside datastore"
+            for info in source_records[source_ref.id]:
+                source_location = info.file_location(source_datastore.locationFactory)
+                target_location = info.file_location(self.locationFactory)
+                if source_location == target_location:
+                    # Either the dataset is already in the target datastore
+                    # (which is how execution butler currently runs) or
+                    # it is an absolute URI.
+                    if source_location.pathInStore.isabs():
+                        # Just because we can see the artifact when running
+                        # the transfer doesn't mean it will be generally
+                        # accessible to a user of this butler. For now warn
+                        # but assume it will be accessible.
```

Comment:
I'm not sure whether this should be an exception or silent, but warning seems like it's still going to be a problem if it's user error, while being annoying if the user knows what they are doing.  Do we want a flag so the user can say, "trust me on this"?

Reply:
In my head this scenario was extremely unlikely because it only happens with ingested datasets and not those created by processing. Since the transfer does not (yet?) copy dimension records over and the receiving registry won't ever have seen these raw files it seems like we aren't really going to get this far in the code. If we turn this into the default mode for transferring content from one butler to another (and only use export/import for cases where you can't see both butlers from your one python process) then yes, we need the code higher up to insert missing dimension records and then ask the user whether absolute URIs should be transferred or left as-is. I was putting in a warning there as a stop-gap scenario where I'm not expecting it to happen but I don't want to raise an exception immediately -- but warning will generally prompt people to ask us what the morning means.